File: code/LD_simulation_GA/LD_stack/Polypred/2_run_SBayesR.py
# Reformat summary statistics and run SBayesR.
# All outputs (including the reformatted summary statistics and a python log file) are in the output_file_dire
# The output result file with posterior effect size is at os.path.join(output_file_dire, output_file_path+".snpRes")

import argparse
import os
import sys
import pandas as pd
import csv

## parse the command line argument
parser = argparse.ArgumentParser(description="Run SBayesR and output the posterior effect sizes.")
parser.add_argument('--race', required=True, type=str, dest="race", 
                    help="the race parameter of summary data")
parser.add_argument('--rho', required=True, type=int, dest="rho", 
                    help="the rho parameter of summary data")
parser.add_argument('--GA', required=True, type=int, dest="GA", 
                    help="the GA parameter of summary data")
parser.add_argument('--size', required=True, type=int, dest="size", 
                    help="the size parameter of summary data")
parser.add_argument('--rep', required=True, type=int, dest="rep", 
                    help="the rep parameter of summary data")
parser.add_argument('--use-public-ldm', default=False, action='store_true', dest="use_public_ldm", 
                    help="If specified, then the program will perform formatting differently")
args = parser.parse_args()

## input parameters
race = args.race
size = args.size
rho = args.rho
GA = args.GA
rep = args.rep

if size == 1:
  N = 15 * 1000
elif size == 2:
  N = 45 * 1000
elif size == 3:
  N = 80 * 1000
else:    
  N = 100 * 1000

## input file path
# summary data (with columns [CHR, POS, SNP_ID, REF, ALT, REF_FRQ PVAL, BETA, SE, N])
# EXAMPLE 1: separate (The following fillers must be included)
is_single_chr_sum = True
sumdata_filepath = "/dcl01/chatterj/data/jin/prs/simulation/{race}/sumdata/megasum-rho{rho}-size{size}-rep{rep}-GA{GA}-chr{chr_num}.txt"
# EXAMPLE 2: combined (absolute filepath with no fillers)
# is_single_chr_sum = False
# sumdata_filepath = "/dcl01/chatterj/data/jin/prs/simulation/{race}/sumdata/megasum-rho{rho}-size{size}-rep{rep}-GA{GA}.txt".format(race=race, rho=rho, size=size, rep=rep, GA=GA) # fake path

## directory of the LD matrix
ld_directory = "/dcs04/nilanjan/data/wlu/PolyPred/SBayesR/ukbEURu_hm3_shrunk_sparse"
ld_mat_file_prefix = "ukbEURu_hm3_{chr_num}_v3_50k.ldm.sparse{file_type}" # (.info / .bin); chr_num = 'all', chr1, chr2
mldm_list_filename = "ukbEURu_hm3_sparse_mldm_list.txt" # will be rewritten with new filepath

## gctb path
gctb_path = "/dcs04/nilanjan/data/wlu/PolyPred/SBayesR/gctb_2.03beta_Linux/gctb"

## output file path
# All outputs (including a python log file) are in the output_file_dire
# The output result file with posterior effect size is at output_file_path+".snpRes"
if args.use_public_ldm:
  print("Use public ldm from {}".format(ld_directory))
output_file_dire = "/dcs04/nilanjan/data/wlu/PolyPred/output/SBayesR_public_ldm/{race}/rho{rho}-size{size}-rep{rep}-GA{GA}".format(race=race,rho=rho,size=size,rep=rep,GA=GA)
else:
  print("Use self ldm from {}".format(ld_directory))
output_file_dire = "/dcs04/nilanjan/data/wlu/PolyPred/output/SBayesR_self_ldm/{race}/rho{rho}-size{size}-rep{rep}-GA{GA}".format(race=race,rho=rho,size=size,rep=rep,GA=GA)

output_file_path = os.path.join(output_file_dire, "SBayesR_result")
python_log = os.path.join(output_file_dire, os.path.basename(__file__[:-3])+".out")


############ Reidrect Output & Sanity Check ###############
# Helper function to run shell command
def run_shell_cmd(shell_cmd):
  print("Running shell command $ " + shell_cmd)
ret = os.popen(shell_cmd)
print(ret.read())

# prep
shell_cmd = "mkdir -p {}".format(output_file_dire)
run_shell_cmd(shell_cmd)

# Redirect stdout
class RedirectOutput:
  
  def __init__(self, filename):
  self.fp = open(filename, 'w')

def write(self, msg):
  self.fp.write(msg)
self.fp.flush()

def flush(self):
  self.fp.flush()

python_out_fp = RedirectOutput(python_log)
sys.stdout = python_out_fp
sys.stderr = python_out_fp

# sanity check
if not os.path.exists(gctb_path):
  msg = "GCTB path" + gctb + " does not exist"
raise Exception(msg)

if is_single_chr_sum:
  for i in range(1, 23):
  curr = sumdata_filepath.format(race=race,rho=rho,size=size,rep=rep,GA=GA,chr_num=i)
if not os.path.exists(curr):
  msg = "Summary stats " + curr + " does not exist"
raise Exception(msg)
else:
  if not os.path.exists(sumdata_filepath):
  msg = "Summary stats " + sumdata_filepath + " does not exist"
raise Exception(msg)

for chr_num in range(1, 23):
  for file_type in [".bin", ".info"]:
  ld_file = ld_mat_file_prefix.format(chr_num="chr"+str(chr_num), file_type=file_type)
ld_path = os.path.join(ld_directory, ld_file)
if not os.path.exists(ld_path):
  msg = "LD matrix file {} does not exist".format(ld_path)
raise Exception(msg)


# ############# Formatting Input Files ################
def read_ld_info(args, path):
  with open(path, 'r') as fp:
  info_lines = fp.readlines()
fp.close()
if not args.use_public_ldm:
  info_lines.pop(0)
return info_lines

print("##################### Formatting Input Files #####################")

all_ld_file = ld_mat_file_prefix.format(chr_num="all", file_type=".info")
all_ld_path = os.path.join(ld_directory, all_ld_file)
if not os.path.exists(all_ld_path):
  print("Create and append to", all_ld_path)
for chr_num in range(1, 23):
  ld_file = ld_mat_file_prefix.format(chr_num="chr"+str(chr_num), file_type=".info")
ld_path = os.path.join(ld_directory, ld_file)
info_lines = read_ld_info(args, ld_path)
with open(all_ld_path, 'a+') as fp:
  for curr_line in info_lines:
  fp.write(curr_line)
if curr_line[-1] != "\n":
  fp.write("\n")
fp.close()
del info_lines

## build the list of LD matrix file paths
mldm_list_path = os.path.join(ld_directory, mldm_list_filename)
with open(mldm_list_path, 'w') as fp:
  for chr_num in range(1, 23):
  ld_file = ld_mat_file_prefix.format(chr_num="chr"+str(chr_num), file_type="")
ld_path = os.path.join(ld_directory, ld_file) + "\n"
fp.write(ld_path)
print("Write the list of LD matrix file paths in {}".format(mldm_list_path))

## format the summary data
# 1. to satisfy SBayesR requirement
# 2. to have consistent SNPID with HapMap3 used for the LD matrix (just in case if this would affect the result)

# sumdata_raw: [CHR, POS, SNP_ID, REF, ALT, REF_FRQ, PVAL, BETA, SE, N, CHR_join, POS_join]
if is_single_chr_sum:
  sumdata_list = []
for i in range(1, 23):
  curr = sumdata_filepath.format(race=race,rho=rho,size=size,rep=rep,GA=GA,chr_num=i)
sumdata_curr = pd.read_table(curr)
sumdata_list.append(sumdata_curr)
sumdata_raw = pd.concat(sumdata_list, axis=0)
else:
  sumdata_raw = pd.read_table(sumdata_filepath)
sumdata_raw["CHR_join"] = sumdata_raw['CHR'].astype(int)
sumdata_raw["POS_join"] = sumdata_raw['POS'].astype(int)
SNP_ID = sumdata_raw["SNP_ID"]
sumdata_raw["SNP_ID"] = [i.strip().split(":")[0].strip() for i in SNP_ID]
del SNP_ID
# sumdata_raw: [CHR, POS, SNP_ID, A2, A1, A1FREQ, PVAL, BETA, SE, N, CHR_join, POS_join]
sumdata_raw = sumdata_raw.rename(columns={"REF": "A2", "ALT": "A1", "REF_FRQ": "A1FREQ"})
sumdata_raw['A1FREQ'] = 1 - sumdata_raw['A1FREQ']
print("Reformat summary data: {} SNPs found in summary data".format(len(sumdata_raw)))

# HapMap info data
# ld_mat_df: ['CHR_join', 'LD_SNP_ID', 'POS_join', "LD_A1", "LD_A2"]
ld_mat_df = pd.read_csv(all_ld_path, sep='\s+', usecols=[0, 1, 3, 4, 5], names=['CHR_join', 'LD_SNP_ID', 'POS_join', "LD_A1", "LD_A2"])
print("Reformat summary data: {} SNPs found in LD matrix".format(len(ld_mat_df)))
ld_mat_df = ld_mat_df.drop(index = ld_mat_df.loc[~ld_mat_df['LD_A1'].isin(['A', 'T', 'G', 'C'])].index)
ld_mat_df = ld_mat_df.drop(index = ld_mat_df.loc[~ld_mat_df['LD_A2'].isin(['A', 'T', 'G', 'C'])].index)
ld_mat_df = ld_mat_df.drop(index = ld_mat_df.loc[ld_mat_df['POS_join'].isna()].index)
ld_mat_df = ld_mat_df.drop(index = ld_mat_df.loc[ld_mat_df['LD_SNP_ID'].isna()].index)
ld_mat_df = ld_mat_df.drop(index = ld_mat_df.loc[ld_mat_df['CHR_join'].isna()].index)
ld_mat_df["CHR_join"] = ld_mat_df['CHR_join'].astype(int)
ld_mat_df["POS_join"] = ld_mat_df['POS_join'].astype(int)
print("Reformat summary data: {} SNPs in LD matrix are valid".format(len(ld_mat_df)))

# perform reformatting
# sumdata_raw: [CHR, POS, SNP_ID, A2, A1, A1FREQ, PVAL, BETA, SE, N, CHR_join, POS_join, LD_SNP_ID, LD_A1, LD_A2]
sumdata_raw = sumdata_raw.merge(ld_mat_df)
del ld_mat_df

print("Reformat summary data: {} SNPs matched with the LD matrix".format(len(sumdata_raw)))
changed_rows = sumdata_raw['SNP_ID']!=sumdata_raw['LD_SNP_ID']
sumdata_raw.loc[changed_rows, 'SNP_ID'] = sumdata_raw.loc[changed_rows, 'LD_SNP_ID']
print("Reformat summary data: {} SNP IDs in the summary data are changed to be consistent with the LD matrix".format(sum(changed_rows)))
del changed_rows

matched_rows = (sumdata_raw['A1']==sumdata_raw['LD_A1']) & (sumdata_raw['A2']==sumdata_raw['LD_A2'])
swapped_rows = (sumdata_raw['A1']==sumdata_raw['LD_A2']) & (sumdata_raw['A2']==sumdata_raw['LD_A1'])
invalid_rows = ~(matched_rows | swapped_rows)
sumdata_raw.loc[swapped_rows, "A1"] = sumdata_raw.loc[swapped_rows, "LD_A1"] 
sumdata_raw.loc[swapped_rows, "A2"] = sumdata_raw.loc[swapped_rows, "LD_A2"] 
sumdata_raw.loc[swapped_rows, "A1FREQ"] = 1 - sumdata_raw.loc[swapped_rows, "A1FREQ"]
sumdata_raw.loc[swapped_rows, "BETA"] = -1 * sumdata_raw.loc[swapped_rows, "BETA"]
sumdata_raw = sumdata_raw.drop(sumdata_raw[invalid_rows].index)
sumdata_raw['Z'] = sumdata_raw['BETA'] / sumdata_raw['SE']
print("Reformat summary data: {} SNPs in summary data were deleted due to unmatched alleles".format(sum(invalid_rows)))
print("Reformat summary data: {} SNPs in summary data were aligned with the LD matrix".format(sum(swapped_rows)))

# save .ma file
sumdata_raw = sumdata_raw[['SNP_ID', 'A1', 'A2', 'A1FREQ', 'BETA', 'SE', 'PVAL', 'N']]
ma_sumstats_file_path = os.path.join(output_file_dire, "sumstats.ma")
sumdata_raw.to_csv(ma_sumstats_file_path, sep=" ", header=True, index=False, doublequote=False, quoting=csv.QUOTE_NONE)

print("Reformat summary data: {} SNPs in summary data remained".format(len(sumdata_raw)))
print("Reformat summary data: stored at {}".format(ma_sumstats_file_path))
print()

del matched_rows, swapped_rows, invalid_rows, sumdata_raw

# ############# Running SBayesR ############
print("##################### Running SBayesR #####################")

final_output_file = output_file_path+"{}.snpRes"
num = 0
gamma_last = 2
while not os.path.exists(final_output_file.format(num)):
  num += 1
gamma_last = gamma_last / 2
if gamma_last < 1e-6:
  break
print("\nNo. {}".format(num))
shell_cmd = """
    {gctb}  --sbayes R \
            --mldm {mldm} \
            --gwas-summary {summary} \
            --out {out} \
            --pi 0.95,0.02,0.02,0.01 \
            --gamma 0,0.01,0.1,{gamma} \
            --seed 0 --chain-length 10000 --burn-in 4000 --out-freq 10 --thin 10 \
             > {log} 2>&1 
    """.format(gctb=gctb_path, 
               mldm=mldm_list_path, 
               summary=ma_sumstats_file_path, 
               out=output_file_path+str(num),
               log=output_file_path+"{}.log".format(num),
               gamma=gamma_last)
run_shell_cmd(shell_cmd)

if not os.path.exists(final_output_file.format(num)):
  print("\nNo. {} (remove the last mixture component)".format(num))
shell_cmd = """
    {gctb}  --sbayes R \
            --mldm {mldm} \
            --gwas-summary {summary} \
            --out {out} \
            --pi 0.95,0.025,0.025 \
            --gamma 0,0.01,0.1 \
            --seed 0 --chain-length 10000 --burn-in 4000 --out-freq 10 --thin 10 \
             > {log} 2>&1 
    """.format(gctb=gctb_path, 
               mldm=mldm_list_path, 
               summary=ma_sumstats_file_path, 
               out=output_file_path+str(num),
               log=output_file_path+"{}.log".format(num))
run_shell_cmd(shell_cmd)

print()
if os.path.exists(final_output_file.format(num)):
  shell_cmd = "cp {orig} {target}".format(orig=final_output_file.format(num), target=final_output_file.format(""))
run_shell_cmd(shell_cmd)
print("Final output", final_output_file.format(""))
else:
  print("Failed")
print("----- END -----")

File: code/LD_simulation_GA/LD_stack/Polypred/2_run_SBayesR.R
# run_XPASS.R runs XPASS and calculate PRS scores (by both XPASS and plink,
# comment the last few lines out if plink is not needed). Note that the way
# to call plink might vary. (It's plink2 --... for me.)
#
# Run with:
# R CMD BATCH --vanill '--args race="AFR" size=1 GA=1 rho=2 rep=1' run_XPASS.R
#
# The section of "input files" controls the input and output file paths.
# Output files have the prefix paste0("/dcs04/nilanjan/data/wlu/XPASS/XPASS_result/", race, "/XPASS-rho",rho,'-size',size,'-rep',rep,'-GA',GA)

## clear workspace
rm(list = ls())
args = commandArgs(trailingOnly = T)
i = as.numeric(args[[1]])
l = as.numeric(args[[2]])
m = as.numeric(args[[3]])
i_rep = as.numeric(args[[4]])
i1 =as.numeric(args[[5]])
library(data.table)
library(tidyverse)
eth <- c("EUR","AFR","AMR","EAS","SAS")
cur.dir <- "/data/zhangh24/multi_ethnic/result/LD_simulation_new/"
out.dir.sum <-  "/data/zhangh24/multi_ethnic/result/LD_simulation_GA/"
out.dir <-  "/data/zhangh24/multi_ethnic/result/LD_simulation_GA/LD_stack/"
sid<-Sys.getenv('SLURM_JOB_ID')
dir.create(paste0('/lscratch/',sid,'/test'),showWarnings = FALSE)
temp.dir = paste0('/lscratch/',sid,'/test/')

setwd("/data/zhangh24/multi_ethnic/")
# input files
summary <- as.data.frame(fread(paste0(out.dir.sum,eth[i],"/summary_mega_rho_",l,"_size_",m,"_rep_",i_rep,"_GA_",i1)))
summary = summary %>% 
  mutate(chr.pos = paste0(CHR, ":", BP))
#match summary data with the reference data of SBayesR
LD_info = fread("/data/zhangh24/software/SBayesR/ukbEURu_hm3_shrunk_sparse/ukbEURu_hm3_all_v3_50k.ldm.sparse.info", header = F)
colnames(LD_info)[1:6] = c("CHR", "RSID", "no1", "BP", "A1", "A2")

LD_info = LD_info %>% select(CHR, RSID, BP, A1, A2) %>% 
  mutate(chr.pos = paste0(CHR, ":", BP)) %>% 
  select(RSID, chr.pos, A1, A2)

#combine summary data with SBayesR
#align the effect to be the same as the reference of SBayesR
#prepare the summary data format to be required format as SBayesR
summary_update = inner_join(summary, LD_info, by = "chr.pos")

summary_update = summary_update %>% 
  mutate(BETA_align = ifelse(effect_allele == A1, BETA, -BETA),
         FQR_align = ifelse(effect_allele == A1, FQR_effect_allele,1 - FQR_effect_allele),
         SE = BETA/Z)

#for(j in 1:22){
  #copy the reference data
  system(paste0("mkdir ",temp.dir,"LD/"))
  LD.dir = paste0(temp.dir,"LD/")
  system(paste0("cp /data/zhangh24/software/SBayesR/ukbEURu_hm3_shrunk_sparse/*.bin ",LD.dir))
  system(paste0("cp /data/zhangh24/software/SBayesR/ukbEURu_hm3_shrunk_sparse/*.info ",LD.dir))
  system(paste0("ls ",LD.dir))
  #create reference file list
  file_name  = rep("c", 22)
  for(j in 1:22){
    file_name[j] = paste0(LD.dir, "ukbEURu_hm3_chr",j,"_v3_50k.ldm.sparse")
    
    
  }
  write.table(file_name, file = paste0(temp.dir, "ukbEURu_hm3_sparse_mldm_list.txt"),
              row.names = F, col.names = F, quote = F)
  mldm_list_filename = paste0(temp.dir, "ukbEURu_hm3_sparse_mldm_list.txt")
  #prepare the summary data format to be required format as SBayesR
  ma = summary_update %>% 
    select(RSID, A1, A2,FQR_align,  BETA_align, SE, P, N) %>% 
    rename(SNP = RSID,
           freq = FQR_align,
           b = BETA_align, 
           se = SE,
           p = P)
  write.table(ma, file = paste0(temp.dir, "sumstats.ma"), row.names = F, col.names = T, quote = F)   
  
  
  
  gctb_path = "/data/zhangh24/software/SBayesR/gctb_2.03beta_Linux/gctb"
  #ref_path = paste0(LD.dir, "ukbEURu_hm3_chr",j,"_v3_50k.ldm.sparse")
  summary_path = paste0(temp.dir, "sumstats.ma")
  out_name =  paste0("/data/zhangh24/multi_ethnic/result/LD_simulation_GA/",eth[i],"/polypred/rho_",l,"_size_",m,"_rep_",i_rep,"_GA_",i1)
  system(paste0(gctb_path," --sbayes R ",
                "--mldm ", mldm_list_filename," ",
                "--pi 0.95,0.02,0.02,0.01 ",
                "--gamma 0.0,0.01,0.1,1 ",
                "--gwas-summary ",summary_path," ",
                "--chain-length 10000 ",
                "--burn-in 4000 ",
                "--out-freq 10 ",
                "--thin 10 ",
                "--out ",out_name))
  #if SBayesR doesn't converge, change the gamma last to gamma_last/2 until convergence
  file_dir = paste0("/data/zhangh24/multi_ethnic/result/LD_simulation_GA/",eth[i],"/polypred")
  files = dir(file_dir, pattern = paste0("rho_",l,"_size_",m,"_rep_",i_rep,"_GA_",i1,".snpRes"),full.names = T)
  out_file = paste0("/data/zhangh24/multi_ethnic/result/LD_simulation_GA/",eth[i],"/polypred/rho_",l,"_size_",m,"_rep_",i_rep,"_GA_",i1,".snpRes")
  gamma_last = 1
  while(out_file%in%files==F){
    gamma_last = gamma_last/2
    system(paste0(gctb_path," --sbayes R ",
                  "--mldm ", mldm_list_filename," ",
                  "--pi 0.95,0.02,0.02,0.01 ",
                  "--gamma 0.0,0.01,0.1,",gamma_last," ",
                  "--gwas-summary ",summary_path," ",
                  "--chain-length 10000 ",
                  "--burn-in 4000 ",
                  "--out-freq 10 ",
                  "--thin 10 ",
                  "--out ",out_name))
    files = dir(file_dir, pattern = paste0("rho_",l,"_size_",m,"_rep_",i_rep,"_GA_",i1),full.names = T)
    if(gamma_last<=1E-06){
      result = NULL
      write.table(result, file = out_file,row.names = F, col.names = F, )
      break
    }
  }
#   
# #}     
#   mis_vec_list = list()   
#   
#   mis_temp =  1
#   
#   for(i in 1:1){
#     file_dir = paste0("/data/zhangh24/multi_ethnic/result/LD_simulation_GA/",eth[i],"/polypred")
#     
#     for(l in 1:3){
#       for(m in 4:4){
#         for(i_rep in 1:10){
#           for(i1 in 1:5){
#             files = dir(file_dir, pattern = paste0("rho_",l,"_size_",m,"_rep_",i_rep,"_GA_",i1,".snpRes"),full.names = T)      
#             out_file = paste0("/data/zhangh24/multi_ethnic/result/LD_simulation_GA/",eth[i],"/polypred/rho_",l,"_size_",m,"_rep_",i_rep,"_GA_",i1,".snpRes")
#             if(out_file%in%files==F){
#               mis_vec_list[[mis_temp]] = data.frame(i, l, m, i_rep, i1)
#               mis_temp = mis_temp + 1
#             }
#             
#           }
#         }
#       }
#     }
#   }
#   
#   for(i in 2:5){
#     file_dir = paste0("/data/zhangh24/multi_ethnic/result/LD_simulation_GA/",eth[i],"/polypred")
#     
#     for(l in 1:3){
#       for(m in 1:4){
#         for(i_rep in 1:10){
#           for(i1 in 1:5){
#             files = dir(file_dir, pattern = paste0("rho_",l,"_size_",m,"_rep_",i_rep,"_GA_",i1,".snpRes"),full.names = T)      
#             out_file = paste0("/data/zhangh24/multi_ethnic/result/LD_simulation_GA/",eth[i],"/polypred/rho_",l,"_size_",m,"_rep_",i_rep,"_GA_",i1,".snpRes")
#             if(out_file%in%files==F){
#               mis_vec_list[[mis_temp]] = data.frame(i, l, m, i_rep, i1)
#               mis_temp = mis_temp + 1
#             }
#             
#           }
#         }
#       }
#     }
#   }
#   
#   mis_vec = rbindlist(mis_vec_list)
#   run_job = rep("c", nrow(mis_vec))
#   for(k in 1:nrow(mis_vec)){
#     run_job[k] = paste0("Rscript /gpfs/gsfs11/users/zhangh24/multi_ethnic/code/LD_simulation_GA/LD_stack/Polypred/2_run_SBayesR.R ",
#                          mis_vec[k,1]," ",mis_vec[k,2]," ",mis_vec[k,3]," ",mis_vec[k,4]," ",mis_vec[k,5])
#   }
#   write.table(run_job, file = paste0("/data/zhangh24/multi_ethnic/code/LD_simulation_GA/LD_stack/Polypred/rerun_SBayesR.sh"),
#               row.names = F, col.names = F, quote = F)
#       